<center> <img src = https://raw.githubusercontent.com/AndreyRysistov/DatasetsForPandas/main/hh%20label.jpg alt="drawing" style="width:400px;">

# <center> Проект: Анализ вакансий из HeadHunter
   

In [1]:
import pandas as pd
import psycopg2

In [2]:
# вставьте сюда параметры подключения из юнита 1. Работа с базой данных из Python 

DBNAME = 'skillfactory'
USER = 'skillfactory'
PASSWORD = 'cCkxxLVrDE8EbvjueeMedPKt'
HOST = '84.201.134.129'
PORT = 5432

In [57]:
# инициализая соединения с базой данных
connection = psycopg2.connect(
    dbname='project_sql',
    user='skillfactory',
    host='84.201.134.129',
    password='cCkxxLVrDE8EbvjueeMedPKt',
    port=5432
)

# Юнит 3. Предварительный анализ данных

1. Напишите запрос, который посчитает количество вакансий в нашей базе (вакансии находятся в таблице vacancies). 

In [4]:
# текст запроса
query_3_1 = f'''
SELECT 
    COUNT(*) Number_of_vacancies
FROM VACANCIES
'''

In [5]:
# результат запроса
df = pd.read_sql_query(query_3_1, connection)
df

C:\Users\roman\AppData\Local\Temp\ipykernel_21096\1114379196.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_1, connection)


number_of_vacancies
0                49197

2. Напишите запрос, который посчитает количество работодателей (таблица employers). 

In [6]:
# текст запроса
query_3_2 = '''
SELECT 
    COUNT(*) Number_of_employers
FROM EMPLOYERS
''' 

In [7]:
# результат запроса
df = pd.read_sql_query(query_3_2, connection)
df

C:\Users\roman\AppData\Local\Temp\ipykernel_21096\3062646138.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_2, connection)


number_of_employers
0                23501

3. Посчитате с помощью запроса количество регионов (таблица areas).

In [8]:
# текст запроса
query_3_3 = '''
SELECT 
   COUNT(*) Number_of_regions
FROM AREAS 
'''

In [9]:
# результат запроса
df = pd.read_sql_query(query_3_3, connection)
df

C:\Users\roman\AppData\Local\Temp\ipykernel_21096\928149064.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_3, connection)


number_of_regions
0               1362

4. Посчитате с помощью запроса количество сфер деятельности в базе (таблица industries).

In [10]:
# текст запроса
query_3_4 = '''
SELECT 
    COUNT(*) 
FROM INDUSTRIES
'''

In [11]:
# результат запроса
df = pd.read_sql_query(query_3_4, connection)
df

C:\Users\roman\AppData\Local\Temp\ipykernel_21096\2101691108.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_4, connection)


count
0    294

***


### <center>**Вывод по предварительному анализу данных** 
Мы имеем достаточно объемную базу данных вакансий, в которой представлены 49197 вакансий,который включают в себя 394 сферы деятельности, 23501 работадатель и 1362 регионов 

# Юнит 4. Детальный анализ вакансий

1. Напишите запрос, который позволит узнать, сколько (cnt) вакансий в каждом регионе (area).
Отсортируйте по количеству вакансий в порядке убывания.

In [12]:
# текст запроса
query_4_1 = """
SELECT 
    a.name Area, 
    COUNT(v.id) Cnt
FROM 
    AREAS a
LEFT JOIN 
    VACANCIES v ON a.id = v.area_id
GROUP BY 
    a.name
ORDER BY 
    cnt DESC
LIMIT 5
"""

In [13]:
# результат запроса
df = pd.read_sql_query(query_4_1, connection)
df

C:\Users\roman\AppData\Local\Temp\ipykernel_21096\2457173557.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_1, connection)


area   cnt
0           Москва  5333
1  Санкт-Петербург  2851
2            Минск  2112
3      Новосибирск  2006
4           Алматы  1892

2. Напишите запрос, чтобы определить у какого количества вакансий заполнено хотя бы одно из двух полей с зарплатой.

In [14]:
# текст запроса
query_4_2 = """
SELECT 
    COUNT(*) Filled_salary_count
FROM 
    VACANCIES
WHERE 
    salary_from IS NOT NULL OR salary_to IS NOT NULL
"""

In [15]:
# результат запроса
df = pd.read_sql_query(query_4_2, connection)
df

C:\Users\roman\AppData\Local\Temp\ipykernel_21096\970655580.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_2, connection)


filled_salary_count
0                24073

3. Найдите средние значения для нижней и верхней границы зарплатной вилки. Округлите значения до целого.

In [16]:
# текст запроса
query_4_3 = '''
SELECT 
    ROUND(AVG(salary_from)) avg_salary_from,
    ROUND(AVG(salary_to)) avg_salary_to
FROM 
    VACANCIES
'''

In [17]:
# результат запроса
df = pd.read_sql_query(query_4_3, connection)
df

C:\Users\roman\AppData\Local\Temp\ipykernel_21096\2500607024.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_3, connection)


avg_salary_from  avg_salary_to
0          71065.0       110537.0

4. Напишите запрос, который выведет количество вакансий для каждого сочетания типа рабочего графика (schedule) и типа трудоустройства (employment), используемого в вакансиях. Результат отсортируйте по убыванию количества.


In [18]:
# текст запроса
query_4_4 = '''
SELECT
    schedule,
    employment,
    COUNT(*) vacancy_count
FROM 
    VACANCIES
GROUP BY
    schedule, employment
ORDER BY 
    vacancy_count DESC
'''

In [19]:
# результат запроса
df = pd.read_sql_query(query_4_4, connection)
df

C:\Users\roman\AppData\Local\Temp\ipykernel_21096\2022935265.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_4, connection)


schedule           employment  vacancy_count
0        Полный день     Полная занятость          35367
1   Удаленная работа     Полная занятость           7802
2      Гибкий график     Полная занятость           1593
3   Удаленная работа  Частичная занятость           1312
4     Сменный график     Полная занятость            940
5        Полный день           Стажировка            569
6     Вахтовый метод     Полная занятость            367
7        Полный день  Частичная занятость            347
8      Гибкий график  Частичная занятость            312
9        Полный день     Проектная работа            141
10  Удаленная работа     Проектная работа            133
11     Гибкий график           Стажировка            116
12    Сменный график  Частичная занятость            101
13  Удаленная работа           Стажировка             64
14     Гибкий график     Проектная работа             18
15    Сменный график           Стажировка             12
16    Вахтовый метод     Проектная работа              2
17    Сменный график     Проектная работа              1

5. Напишите запрос, выводящий значения поля Требуемый опыт работы (experience) в порядке возрастания количества вакансий, в которых указан данный вариант опыта. 

In [20]:
# текст запроса
query_4_5 = '''
SELECT
    experience,
    COUNT(*) vacancy_count
FROM 
    VACANCIES
GROUP BY
    experience
ORDER BY 
    vacancy_count
'''

In [21]:
# результат запроса
df = pd.read_sql_query(query_4_5, connection)
df

C:\Users\roman\AppData\Local\Temp\ipykernel_21096\2316260449.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_5, connection)


experience  vacancy_count
0         Более 6 лет           1337
1           Нет опыта           7197
2       От 3 до 6 лет          14511
3  От 1 года до 3 лет          26152

***

### <center> **Вывод по детальному анализу данных**
Больше всего вакансий приходится на крупнейшией по численности населения региона России, Белоруси и Казахстана (два города столицы) - Москва, Санкт-Петербург, Минск, Новосибирск, Алматы. Чуть меньше половины вакансий содержат как верхнюю, так и нижнюю границы зарплаты. Бесспорный лидер, что касается типа рабочего графика, это полный день в сочетании с полной занятостью, что, очевидно, говорит и оживлении эномоки регионов после отмены короновирусных ограничений.Наибольшее количество вакансий содержат требование, чтобы соискатель уже имел некоторый опыт работы, от 1 года до 3 лет. Вакансий же без опыта работы в два раза меньше. Меньше всего вакансий в базе данных, которые требуют от соискателей опыта работа более 6 лет и кандидатов без опыта - это свясано с их узкой специализацией и особенностями найма. 


# Юнит 5. Анализ работодателей

1. Напишите запрос, который позволит узнать, какие работодатели находятся на первом и пятом месте по количеству вакансий.

In [22]:
# текст запроса
query_5_1 = '''
SELECT 
    e.name Name_of_employer, 
    COUNT(v.id) Cnt
FROM 
    EMPLOYERS e
LEFT JOIN 
    VACANCIES v ON e.id = v.employer_id
GROUP BY 
    e.name
ORDER BY 
    cnt DESC
LIMIT 5
'''

In [23]:
# результат запроса
df = pd.read_sql_query(query_5_1, connection)
df

C:\Users\roman\AppData\Local\Temp\ipykernel_21096\3834399874.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_1, connection)


name_of_employer   cnt
0           Яндекс  1933
1       Ростелеком   491
2         Тинькофф   444
3             СБЕР   428
4    Газпром нефть   331

2. Напишите запрос, который для каждого региона выведет количество работодателей и вакансий в нём.
Среди регионов, в которых нет вакансий, найдите тот, в котором наибольшее количество работодателей.


In [24]:
# текст запроса
query_5_2 = '''
SELECT 
    a.name region_name, 
    COUNT(DISTINCT e.id) employer_count
FROM 
    AREAS a
LEFT JOIN 
    EMPLOYERS e ON a.id = e.area
LEFT JOIN 
    VACANCIES v ON a.id = v.area_id
WHERE 
    v.id IS NULL
GROUP BY 
    a.name
ORDER BY 
    employer_count DESC
LIMIT 1
'''

In [25]:
# результат запроса
df = pd.read_sql_query(query_5_2, connection)
df

C:\Users\roman\AppData\Local\Temp\ipykernel_21096\2250345976.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_2, connection)


region_name  employer_count
0      Россия             410

3. Для каждого работодателя посчитайте количество регионов, в которых он публикует свои вакансии. Отсортируйте результат по убыванию количества.


In [26]:
# текст запроса
query_5_3 = '''
SELECT 
    employer_id,
    COUNT(DISTINCT area_id) region_count
FROM 
    vacancies
GROUP BY 
    employer_id
ORDER BY 
    region_count DESC
'''

In [27]:
# результат запроса
df = pd.read_sql_query(query_5_3, connection)
df

C:\Users\roman\AppData\Local\Temp\ipykernel_21096\201667002.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_3, connection)


employer_id  region_count
0             1740           181
1             2748           152
2          5724811           116
3          5130287            88
4          3682876            71
...            ...           ...
14901       810278             1
14902       810313             1
14903       810551             1
14904       810688             1
14905       811902             1

[14906 rows x 2 columns]

4. Напишите запрос для подсчёта количества работодателей, у которых не указана сфера деятельности. 

In [28]:
# текст запроса
query_5_4 = '''
SELECT 
    COUNT(*) cnt
FROM 
    EMPLOYERS e
LEFT JOIN 
    EMPLOYERS_INDUSTRIES ei on e.id = ei.employer_id
WHERE 
    ei.industry_id is null
'''

In [29]:
# результат запроса
df = pd.read_sql_query(query_5_4, connection)
df

C:\Users\roman\AppData\Local\Temp\ipykernel_21096\3493105328.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_4, connection)


cnt
0  8419

5. Напишите запрос, чтобы узнать название компании, находящейся на третьем месте в алфавитном списке (по названию) компаний, у которых указано четыре сферы деятельности. 

In [30]:
# текст запроса
query_5_5 = '''
SELECT 
    e.name 
FROM 
    EMPLOYERS e
LEFT JOIN 
    EMPLOYERS_INDUSTRIES ei ON e.id = ei.employer_id
GROUP BY 
    e.id
HAVING
    COUNT(ei.industry_id) = 4
ORDER BY 
    e.name
OFFSET 2 LIMIT 1
'''

In [31]:
# результат запроса
df = pd.read_sql_query(query_5_5, connection)
df

C:\Users\roman\AppData\Local\Temp\ipykernel_21096\388711819.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_5, connection)


name
0  2ГИС

6. С помощью запроса выясните, у какого количества работодателей в качестве сферы деятельности указана Разработка программного обеспечения.


In [32]:
# текст запроса
query_5_6 = '''
SELECT 
    COUNT(DISTINCT e.id) software_development_employers
FROM 
    employers e
JOIN 
    employers_industries ei ON e.id = ei.employer_id
JOIN 
    industries i ON ei.industry_id = i.id
WHERE 
    i.name = 'Разработка программного обеспечения'
'''

In [33]:
# результат запроса
df = pd.read_sql_query(query_5_6, connection)
df

C:\Users\roman\AppData\Local\Temp\ipykernel_21096\821628714.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_6, connection)


software_development_employers
0                            3553

7. Для компании «Яндекс» выведите список регионов-миллионников, в которых представлены вакансии компании, вместе с количеством вакансий в этих регионах. Также добавьте строку Total с общим количеством вакансий компании. Результат отсортируйте по возрастанию количества.

Список городов-милионников надо взять [отсюда](https://ru.wikipedia.org/wiki/%D0%93%D0%BE%D1%80%D0%BE%D0%B4%D0%B0-%D0%BC%D0%B8%D0%BB%D0%BB%D0%B8%D0%BE%D0%BD%D0%B5%D1%80%D1%8B_%D0%A0%D0%BE%D1%81%D1%81%D0%B8%D0%B8). 

Если возникнут трудности с этим задание посмотрите материалы модуля  PYTHON-17. Как получать данные из веб-источников и API. 

In [ ]:
from urllib.parse import quote_plus
import sqlite3 

# Получение списка городов-миллионеров с Wikipedia
url = quote_plus('https://ru.wikipedia.org/wiki/Города-миллионеры_России', safe=':/')
tables = pd.read_html(url)
mln_cities = tuple(tables[0]['Город'])

# Преобразование списка в строку для SQL
mln_cities_sql = ', '.join(f"'{city}'" for city in mln_cities)  

query_5_7 = f"""
SELECT 
    a.name AS region,
    COUNT(v.id) AS vacancies_count
FROM 
    vacancies AS v
JOIN employers AS e ON v.employer_id = e.id
JOIN areas AS a ON v.area_id = a.id
WHERE 
    e.name = 'Яндекс' AND a.name IN ({mln_cities_sql})
GROUP BY 
    a.name

UNION ALL

SELECT 
    'Total' AS region,
    COUNT(v.id) AS vacancies_count
FROM 
    vacancies AS v
JOIN employers AS e ON v.employer_id = e.id
JOIN areas AS a ON v.area_id = a.id
WHERE 
    e.name = 'Яндекс' AND a.name IN ({mln_cities_sql})
ORDER BY 
    vacancies_count DESC
"""
df = pd.read_sql_query(query_5_7, connection)

print(df)

C:\Users\roman\AppData\Local\Temp\ipykernel_21096\2933541463.py:38: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_7, connection)


             region  vacancies_count
0             Total              485
1            Москва               54
2   Санкт-Петербург               42
3      Екатеринбург               39
4   Нижний Новгород               36
5       Новосибирск               35
6           Воронеж               32
7         Краснодар               30
8            Самара               26
9               Уфа               26
10            Пермь               25
11           Казань               25
12   Ростов-на-Дону               25
13        Волгоград               24
14       Красноярск               23
15        Челябинск               22
16             Омск               21


***

### <center> **Выводы по анализу работодателей**
На первом месте по количеству вакансий находится компания "Яндекс", так как имеет целую группу дочерних компаний в различных сферах деятельности: Яндекс.Такси, Яндекс Бизнес, Яндекс Драйв, Яндекс Маркет и тд. Почти четверть работадателей имеют сферу деятельности "Разработка программного обеспечения". Наибольшее количество работадателей находится на территории России  


# Юнит 6. Предметный анализ

1. Сколько вакансий имеет отношение к данным?

Считаем, что вакансия имеет отношение к данным, если в её названии содержатся слова 'data' или 'данн'.

*Подсказка: Обратите внимание, что названия вакансий могут быть написаны в любом регистре.* 


In [35]:
# текст запроса
query_6_1 = """
SELECT 
    COUNT(*)
FROM 
    VACANCIES
WHERE 
    LOWER(name) LIKE '%data%' OR LOWER(name) LIKE '%данн%'
"""

In [36]:
# результат запроса
df = pd.read_sql_query(query_6_1, connection)
df

C:\Users\roman\AppData\Local\Temp\ipykernel_21096\366702258.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_1, connection)


count
0   1771

2. Сколько есть подходящих вакансий для начинающего дата-сайентиста? 
Будем считать вакансиями для дата-сайентистов такие, в названии которых есть хотя бы одно из следующих сочетаний:
* 'data scientist'
* 'data science'
* 'исследователь данных'
* 'ML' (здесь не нужно брать вакансии по HTML)
* 'machine learning'
* 'машинн%обучен%'

** В следующих заданиях мы продолжим работать с вакансиями по этому условию.*

Считаем вакансиями для специалистов уровня Junior следующие:
* в названии есть слово 'junior' *или*
* требуемый опыт — Нет опыта *или*
* тип трудоустройства — Стажировка.
 

In [37]:
# текст запроса
query_6_2 = '''
SELECT
    COUNT(v.name)
FROM VACANCIES v
WHERE
(
    (name ILIKE '%data scientist%') OR (name ILIKE '%data science%') OR (name ILIKE '%исследователь данных%')
    OR ((name ILIKE '%ML%') AND (name NOT ILIKE '%HTML%'))
    OR (name ILIKE '%machine learning%')
    OR (name ILIKE '%машинн%обучен%')
    AND (key_skills IS NOT NULL)
)

    AND (v.name ILIKE '%junior%'
    OR v.experience ILIKE '%Нет опыта%'
    OR v.employment ILIKE '%Стажировка%')
'''

In [38]:
# результат запроса
df = pd.read_sql_query(query_6_2, connection)
df

C:\Users\roman\AppData\Local\Temp\ipykernel_21096\2267038862.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_2, connection)


count
0     51

3. Сколько есть вакансий для DS, в которых в качестве ключевого навыка указан SQL или postgres?

** Критерии для отнесения вакансии к DS указаны в предыдущем задании.*

In [39]:
# текст запроса
query_6_3 = '''
SELECT
    v.id, v.key_skills
FROM vacancies v
WHERE ( (v.key_skills like '%SQL%')
        or
        (v.key_skills ilike '%postgres%')
        )
    AND
(
    (name ILIKE '%data scientist%') OR (name ILIKE '%data science%') OR (name ILIKE '%исследователь данных%')
    OR ((name ILIKE '%ML%') AND (name NOT ILIKE '%HTML%'))
    OR (name ILIKE '%machine learning%')
    OR (name ILIKE '%машинн%обучен%')
    AND (key_skills IS NOT NULL)
)
ORDER BY 1
'''

In [40]:
# результат запроса
df = pd.read_sql_query(query_6_3, connection)
df

C:\Users\roman\AppData\Local\Temp\ipykernel_21096\3794299702.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_3, connection)


id                                         key_skills
0    48838707  Git\tSpring Framework\tJava\tPostgreSQL\tApach...
1    49930707  Linux\tPostgreSQL\tSQL\tPython\tJava\tPHP\tPer...
2    50533416          Python\tSQL\tРабота в команде\tOpenCV\tML
3    51001456          Python\tGit\tLinux\tSQL\tBig Data\tGolang
4    51834014  JavaScript\tMySQL\tОбучение и развитие\tУправл...
..        ...                                                ...
224  66134973  MS PowerPoint\tPython\tSQL\tПодготовка презент...
225  66142163  SQL\tMS SQL\tУдаленная работа\tPython\tМатемат...
226  66176514  Python\tJavaScript\tDjango Framework\tReact\tM...
227  66178607  Python\tPostgreSQL\tPyTorch\tTensorflow\tPanda...
228  66185427  Python\tLinux\tАнализ данных\tKeras\tData Scie...

[229 rows x 2 columns]

4. Проверьте, насколько популярен Python в требованиях работодателей к DS.Для этого вычислите количество вакансий, в которых в качестве ключевого навыка указан Python.

** Это можно сделать помощью запроса, аналогичного предыдущему.*

In [41]:
# текст запроса
query_6_4 = '''
SELECT
    v.id, v.key_skills
FROM
    vacancies v
WHERE ( (v.key_skills like '%Python%')
        OR
        (v.key_skills ilike '%python%')
        )
    AND
(
    (name ILIKE '%data scientist%') OR (name ILIKE '%data science%') OR (name ILIKE '%исследователь данных%')
    OR ((name ILIKE '%ML%') AND (name NOT ILIKE '%HTML%'))
    OR (name ILIKE '%machine learning%')
    OR (name ILIKE '%машинн%обучен%')
    AND (key_skills IS NOT NULL)
)
ORDER BY 1 
'''

In [42]:
# результат запроса
df = pd.read_sql_query(query_6_4, connection)
df

C:\Users\roman\AppData\Local\Temp\ipykernel_21096\4264665179.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_4, connection)


id                                         key_skills
0    43183814  Python\tC++\tНаписание статей\tНаписание научн...
1    43366306  Python\tРиск-менеджмент\tМатематическая статис...
2    44095047         Python\tGit\tAtlassian Jira\tLinux\tMATLAB
3    47094497       SCALA\tSpark\tHadoop\tJava\tBig Data\tPython
4    47095350       SCALA\tSpark\tHadoop\tJava\tBig Data\tPython
..        ...                                                ...
352  66142163  SQL\tMS SQL\tУдаленная работа\tPython\tМатемат...
353  66146489  Python\tPyTorch\tOpenCV\tComputer Vision\tTens...
354  66176514  Python\tJavaScript\tDjango Framework\tReact\tM...
355  66178607  Python\tPostgreSQL\tPyTorch\tTensorflow\tPanda...
356  66185427  Python\tLinux\tАнализ данных\tKeras\tData Scie...

[357 rows x 2 columns]

5. Сколько ключевых навыков в среднем указывают в вакансиях для DS?
Ответ округлите до двух знаков после точки-разделителя.

In [43]:
# текст запроса
query_6_5 = '''
SELECT
    ROUND(AVG(LENGTH(key_skills) - LENGTH(REPLACE(key_skills,CHR(9),''))+1),2)
FROM VACANCIES v
WHERE
(
       (v.name ILIKE '%data scientist%')
    OR (lower(v.name) ILIKE '%data science%')
    OR (v.name ILIKE '%исследователь данных%')
    OR (v.name  LIKE '%ML%' AND  (v.name NOT iLIKE '%html%'))
    OR (v.name ILIKE '%machine learning%')
    OR (v.name ILIKE '%машинн%обучен%')
    AND (key_skills IS NOT NULL)
)
'''

In [44]:
# результат запроса
df = pd.read_sql_query(query_6_5, connection)
df

C:\Users\roman\AppData\Local\Temp\ipykernel_21096\3654210517.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_5, connection)


round
0   6.41

6. Напишите запрос, позволяющий вычислить, какую зарплату для DS в **среднем** указывают для каждого типа требуемого опыта (уникальное значение из поля *experience*). 

При решении задачи примите во внимание следующее:
1. Рассматриваем только вакансии, у которых заполнено хотя бы одно из двух полей с зарплатой.
2. Если заполнены оба поля с зарплатой, то считаем зарплату по каждой вакансии как сумму двух полей, делённую на 2. Если заполнено только одно из полей, то его и считаем зарплатой по вакансии.
3. Если в расчётах участвует null, в результате он тоже даст null (посмотрите, что возвращает запрос select 1 + null). Чтобы избежать этой ситуацию, мы воспользуемся функцией [coalesce](https://postgrespro.ru/docs/postgresql/9.5/functions-conditional#functions-coalesce-nvl-ifnull), которая заменит null на значение, которое мы передадим. Например, посмотрите, что возвращает запрос `select 1 + coalesce(null, 0)`

Выясните, на какую зарплату в среднем может рассчитывать дата-сайентист с опытом работы от 3 до 6 лет. Результат округлите до целого числа. 

In [45]:
# текст запроса
query_6_6 = '''
SELECT
    v.experience, ROUND(AVG(COALESCE((v.salary_from + v.salary_to) / 2, v.salary_from, v.salary_to)), 0)
FROM 
    VACANCIES v

WHERE
(
       (v.name ILIKE '%data scientist%')
    OR (lower(v.name) ILIKE '%data science%')
    OR (v.name ILIKE '%исследователь данных%')
    OR (v.name  LIKE '%ML%' AND  (v.name NOT iLIKE '%html%'))
    OR (v.name ILIKE '%machine learning%')
    OR (v.name ILIKE '%машинн%обучен%')
    AND (key_skills IS NOT NULL)
)
GROUP BY 
    v.experience
ORDER BY 1
'''

In [46]:
# результат запроса
df = pd.read_sql_query(query_6_6, connection)
df

C:\Users\roman\AppData\Local\Temp\ipykernel_21096\519194770.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_6, connection)


experience     round
0         Более 6 лет       NaN
1           Нет опыта   74643.0
2  От 1 года до 3 лет  145236.0
3       От 3 до 6 лет  243115.0

***

### <center> **Выводы по предметному анализу**
1771 вакансия, имеющая отношение к данным, свидетельствует о том, что данные вакансии пользуются некоторым спросом, однако в общей массе не занимают большую долю. 51 вакансия для начинающего дата-сайентиста это очень маленькое число от общего количества вакансий, связанных с данными. Это говорит о том, что позиции для Junior-уровня в сфере Data Science ограничены, что может быть вызвано высокой конкуренцией и предпочтением компаний к более опытным специалистам. Python оказался более востребованным и популярным навыком, чем SQL или Postgres. Среднее количество ключевых навыков со значением 6.41, свидетельствует о высоком показателе, который подчеркивает, что работодатели ожидают от дата-сайентиста многогранного набора компетенций

**Сравнение "зараплата" vs "опыт"**

* **Нет опыта: 74,643 руб.** Это начальная зарплата, которая может быть привлекательной для начинающих специалистов, но в то же время отражает ограниченные возможности роста для Junior на старте.
* **От 1 года до 3 лет: 145,236 руб.** Значительное увеличение по сравнению с "Нет опыта", что указывает на рост ценности специалиста даже при минимальном опыте.
* **От 3 до 6 лет: 243,115 руб.** Это подтверждает, что опыт играет ключевую роль в определении заработной платы.
* **Более 6 лет: NaN.** Это может быть связано с отсутствием достаточного количества данных для вакансий такого уровня или с трудностью их классификации.

## <center> **Общий вывод по проекту**

Больше всего вакансий расположено в крупнейших городах России, Белоруси и Казахстана, крупнейших как по численности населения и как центрах технологических корпораций. Data Science остается достаточно востребованной областью, особенно для специалистов с опытом. Начинающим специалистам приходится конкурировать за небольшое число позиций. Зарплата в сфере Data Science сильно зависит от опыта, что стимулирует профессиональный рост и развитие навыков. Средний уровень зарплаты показывает высокую ценность специалистов этой сферы. Работодатели ожидают разнообразных навыков, что требует от кандидатов быть хорошо подготовленными и универсальными.


# ЮНИТ 7. Дополнительное исследование

**Вывод по дополнительному исследованию:** в связи с тем, что в базе данных только 51 вакансия для начинающего дата-сайентиста от общего количества вакансий, то я захотел помочь новичкам с тем, где отзываться на вакансии, то есть в каких регионах и в каким компаниях. Безусловными лидерами являются Москва и Банк ВТБ

In [ ]:
# ТОП-10 вакансий
query_6_7_add = """
SELECT 
    name vacancy_name, 
    COUNT(*) vacancy_count
FROM 
    vacancies
GROUP BY 
    name
ORDER BY 
    vacancy_count DESC
LIMIT 10
"""

In [48]:
# результат запроса
df = pd.read_sql_query(query_6_7_add, connection)
df

C:\Users\roman\AppData\Local\Temp\ipykernel_21096\279163513.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_7_add, connection)


vacancy_name  vacancy_count
0           Системный администратор           1761
1                    Программист 1С           1153
2                   Бизнес-аналитик            488
3  Специалист технической поддержки            457
4                Системный аналитик            432
5                          Аналитик            404
6               Инженер-программист            389
7                    Программист 1C            360
8              Frontend-разработчик            242
9     Инженер технической поддержки            230

In [ ]:
#ТОП-10 регинов, в которых предлагают работу Junior'am
query_6_8_add = '''
SELECT 
    a.name region, 
    COUNT(v.id) junior_vacancies_count
FROM 
    vacancies v
JOIN 
    areas a ON v.area_id = a.id
WHERE
    (
        (v.name ILIKE '%data scientist%') OR (v.name ILIKE '%data science%') 
        OR (v.name ILIKE '%исследователь данных%')
        OR ((v.name ILIKE '%ML%') AND (v.name NOT ILIKE '%HTML%'))
        OR (v.name ILIKE '%machine learning%')
        OR (v.name ILIKE '%машинн%обучен%')
    )
    AND (v.name ILIKE '%junior%' 
         OR v.experience ILIKE '%Нет опыта%' 
         OR v.employment ILIKE '%Стажировка%')
GROUP BY 
    a.name
ORDER BY 
    junior_vacancies_count DESC
LIMIT 10
'''

In [54]:
df = pd.read_sql_query(query_6_8_add, connection)
df

C:\Users\roman\AppData\Local\Temp\ipykernel_21096\943791187.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_8_add, connection)


region  junior_vacancies_count
0           Москва                      19
1  Санкт-Петербург                       9
2      Новосибирск                       5
3            Минск                       2
4          Воронеж                       2
5     Екатеринбург                       2
6   Ростов-на-Дону                       2
7            Пермь                       2
8            Томск                       2
9           Казань                       2

In [ ]:
#ТОП-10 компаний, в которых предлагают работу Junior'am
query_6_9_add = '''
SELECT 
    e.name AS company_name, 
    COUNT(v.id) junior_vacancies_count
FROM 
    vacancies v
JOIN 
    employers e ON v.employer_id = e.id
WHERE
    (
        (v.name ILIKE '%data scientist%') OR (v.name ILIKE '%data science%') 
        OR (v.name ILIKE '%исследователь данных%')
        OR ((v.name ILIKE '%ML%') AND (v.name NOT ILIKE '%HTML%'))
        OR (v.name ILIKE '%machine learning%')
        OR (v.name ILIKE '%машинн%обучен%')
    )
    AND (v.name ILIKE '%junior%' 
         OR v.experience ILIKE '%Нет опыта%' 
         OR v.employment ILIKE '%Стажировка%')
GROUP BY 
    e.name
ORDER BY 
    junior_vacancies_count DESC
LIMIT 10
'''

In [52]:
df = pd.read_sql_query(query_6_9_add, connection)
df

C:\Users\roman\AppData\Local\Temp\ipykernel_21096\3917716556.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_9_add, connection)


company_name  junior_vacancies_count
0                                    Банк ВТБ (ПАО)                      10
1                                            Comexp                       3
2  The Skolkovo Institute of Science and Technology                       3
3                                     SteadyControl                       2
4     СтарЛайн, Научно-производственное объединение                       2
5                                            Диплей                       2
6                                            Seldon                       1
7                                 DNS Головной офис                       1
8                                                VK                       1
9                                          Pay P.S.                       1